azerothcore/azerothcore-wotlk
# feat(Core/SmartAI): Implement AoE variants of VICTIM_CASTING and RANGE events

## Changes Proposed:
- Adds a new action type to allow creatures to play visualkits without needing to cast spells
- Adds NEAR_UNIT_NEGATION counterpart
- Adjusts logic to NEAR_PLAYER_NEGATION
- Implements SMART_ACTION_AREA_RANGE to search if there are any players nearby on range. This should be used for spells like Blast Wave or Frozen Nova mainly. Difference between NEAR_PLAYER/NEAR_UNIT is that this is In Combat only.

This is changes a lot of mobs previously using incorrect casting events. For example, casting Frost Nova when no one's nearby

Notes:
ALL events that are used for spells should have 4 parameters reserved for initial AND repeat timers
![image](https://github.com/azerothcore/azerothcore-wotlk/assets/100873791/c08c9913-9d48-43db-a6db-c74ba53692ea)
Both new events recalculate timer if spell is failed or no targets are found. This is not purely speculation. For example, when Cleave spells fail, they are attempted again 1200ms, which is 3 packet cycles or every Blizz creature AI update

## Issues Addressed:

- Closes https://github.com/chromiecraft/chromiecraft/issues/5705
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/13934
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/12301
- Closes https://github.com/chromiecraft/chromiecraft/issues/3653

May close more but I did not check further

## SOURCE:


## Tests Performed:

- Somewhat tested
- 


## How to Test the Changes:


```
DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` = 448);
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(448, 0, 0, 0, 38, 0, 100, 0, 1, 1, 0, 0, 0, 229, 90, 0, 0, 0, 0, 0, 9, 0, 0, 100, 0, 0, 0, 0, 0, 'Hogger - On Data Set - Play Spell Visual Kit'),
(448, 0, 1, 0, 105, 0, 100, 0, 3000, 3000, 3000, 3000, 30, 11, 29443, 66, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Hostile Casting in 30y - Cast \'Counterspell\''),
(448, 0, 2, 0, 106, 0, 100, 0, 3000, 3000, 3000, 3000, 10, 11, 11831, 66, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Hostile in Range 10y - Cast \'Frost Nova\'');
```
`.go c id 448`
`.npc set data 1 1`
see that every creature within 100y plays a visual

Get within 30y and try to cast
Get within 10y to see frost nova

## Known Issues and TODO List:




## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
